### ISSUE TITLE
geoip2 critical error (getWithPrefixLen)

### ISSUE THREAD

**Issue by ideenmacher** [wp-statistics/wp-statistics#439]
Dear Team,

any idea why this error still can occur? I have the newest Update installed.

according to the logs it only happens every few hours, which may mean that some user (or bot) origins may not be detected correctly? how to solve this?

CRITICAL Uncaught Error: Call to undefined method MaxMind\Db\Reader::getWithPrefixLen() in … /plugins/wp-statistics/includes/vendor/geoip2/geoip2/src/Database/Reader.php:246
Stack trace:
0 … /plugins/wp-statistics/includes/vendor/geoip2/geoip2/src/Database/Reader.php(217): GeoIp2\Database\Reader->getRecord()
1 … /plugins/wp-statistics/includes/vendor/geoip2/geoip2/src/Database/Reader.php(90): GeoIp2\Database\Reader->modelFor()
2 … /plugins/wp-statistics/includes/class-wp-statistics-geoip.php(183): GeoIp2\Database\Reader->country()
3 … /plugins/wp-statistics/includes/class-wp-statistics-exclusion.php(399): WP_STATISTICS\GeoIP::getCountry()
4 …/plugins/wp-statistics/includes/class-wp-statistics-exclusion.php(63): in … /plugins/wp-statistics/includes/vendor/geoip2/geoip2/src/Database/Reader.php in row 246

Thanks!

**Comment by mostafasoufi** [comment]
You're correct, I've just fixed it on the development version, can you please test it? https://github.com/wp-statistics/wp-statistics/commit/daf5b219650219444c9f510c6bc47cc86f3416bc

Cheers!

**Comment by mostafasoufi** [comment]
Let's replace all files since we made a couple of improvements on the previous commits.

>And you mean testing it with country/city option enabled or disabled?

The current situation.

Also, there is no rush, let's wait and test with some real visitors.

Thank you!

**Comment by mostafasoufi** [comment]
Just download the [development](https://github.com/wp-statistics/wp-statistics/archive/refs/heads/development.zip) branch and then replace Its file on the server.

> Can a disabled Online-User-Tracking option make a difference?

Yeah, you can easily disable it in **Statistics → Settings → General**

**Comment by mostafasoufi** [comment]
Weird indeed, please make sure the below files are available on your server:

```
wp-content/uploads/wp-statistics/GeoLite2-City.mmdb
wp-content/uploads/wp-statistics/GeoLite2-Country.mmdb
```

If not, you can download it manually from https://github.com/wp-statistics/GeoLite2-Country and https://github.com/wp-statistics/GeoLite2-City

**Comment by mostafasoufi** [comment]
Thank you for testing and your effort. Your efforts are appreciated.

You don't need to clear the cache.

Let's do some Composer autoloading tests on the site. Do you have the Composer CLI installed on your server? If so, let's try running `composer update` in the plugin path directory `wp-content/plugins/wp-statistics`.

**Comment by mostafasoufi** [comment]
> these error logs: Duplicate entry 'XX' for key 'page_id' are occurring 20-30 times at the same second with different ids. that might be something, or?

Not really.

> what are the PHP version requirements of the plugin and the maxmind library?

Would be good to use PHP v7.4 ~ PHP v8.0

> ad news: also i saw, that short statistic now shows 0 visitors/visits for today, which definitely is not correct. maybe the conditional functionality prevents the statistic hits to be triggered, when geoip options are disabled?)

Wierd indeed, will review again.

**Comment by mostafasoufi** [comment]
We've look it on the issue till Monday and will get back to you soon.

**Comment by oschwald** [comment]
Based on the error, I suspect the host has a version of the `maxminddb` C extension that is older than 1.5.0 (released in 2019). This is not compatible with newer versions of the GeoIP2 package as it doesn't provide the necessary lookup method.

**Comment by mostafasoufi** [comment]
Good point, I've updated `maxmind/web-service-common` v0.7.0 to v0.9.0 and `maxmind-db/reader` v1.6.0 => v1.11.0, so can you please test it?

https://www.dropbox.com/scl/fi/jqe3px3aye6xh4bpv5vlj/wp-statistics.zip?rlkey=7zibeawizdnqb5sduv3yupzug&dl=0

**Comment by oschwald** [comment]
Sorry, I don't think I phrased my comment well. I was trying to say that the host likely has a very old version of the [`maxminddb` PECL extension](https://pecl.php.net/package/maxminddb) installed. You could confirm this with `get_loaded_extensions()` from PHP or `php -m` from the command line (but be sure it is using the same config that it would use for WP). You will likely have to ask your hosting provider to update the extension.

If this hypothesis is not correct, then I would assume some other WordPress plugin is loading a very old version of `MaxMind\Db`, preventing this plugin from loading the newer version required by the GeoIP2 package.

In terms of what WP Statistics could do to prevent users from hitting either of these issues, one solution would be to use a tool like [PHP-Scoper](https://github.com/humbug/php-scoper) to namespace all dependencies, preventing collisions with other plugins or installed extensions. This would have the disadvantage of users not getting the performance benefit of the C extension if it is installed. (I don't know how many lookups this plugin does so it is hard for me to say whether that would be a significant issue.)

**Comment by alifallahrn** [comment]
Hey @ideenmacher 

To investigate potential conflicts with plugins, we need precise information on the plugins and their versions installed on your website.
So, please follow these steps to share the necessary data:

1. Log in to your WordPress admin dashboard.
2. Go to "Tools" -> "Site Health."
3. Navigate to the "Info" tab.
4. Click "Copy site info to clipboard" and send us the data.

Overall, we have concluded that we should fix this issue with the PHP Scoper solution to ensure that we won't have any conflicts with other plugins in the future.

---

@oschwald Thanks for providing the PHP Scoper solution. Yes, we also believe it's the best solution. We use this library in our other products, and it was in our plans to add it to WP Statistics in the future, but it seems like it's time to go ahead with it now.

**Comment by mostafasoufi** [comment]
Soon!

Let's mention the issue number `... #439` in the commits message that is related to this issue. in this case, all will be notified about any changes. @alifallahrn

**Comment by mostafasoufi** [comment]
Sounds good, thank you for testing! we'll review the error asap. @ideenmacher

**Comment by mostafasoufi** [comment]
Btw, for testing, you need to build the plugin, but you need to install the PHP-Scoper globally in advance.

```
composer global require humbug/php-scoper "0.16.1"
```

Then, build it.

```
composer build-zip --ignore-platform-reqs
```

**Comment by mostafasoufi** [comment]
I'll send you the build version soon.

**Comment by mostafasoufi** [comment]
Update: Seems the build version has some issues, trying to fix it and will get back to you guys.

**Comment by mostafasoufi** [comment]
We had trouble organizing the classes, and some urgent tasks came up. I'll work on it and let you know once it's sorted.

**Comment by mostafasoufi** [comment]
Sorry for the late, can you please test this scoped version? https://www.dropbox.com/scl/fi/6kizt2g1osgo0f9oeo8dg/wp-statistics.zip?rlkey=lz12sgdzdz4zh86bn76vk625i&dl=0

**Comment by mostafasoufi** [comment]
Resolving the issue is certainly possible, but it requires a considerable amount of time. This is because we need to test each part thoroughly upon completion, which extends the process somewhat. Therefore, we might need to address this with a bit of a delay and return to this issue later on.

**Comment by mostafasoufi** [comment]
FYI: Good news, seems there is a new solution called [Mozart](https://github.com/coenjacobs/mozart) which this [plugin](https://github.com/pressidium/pressidium-cookie-consent) ([Built version](https://plugins.trac.wordpress.org/browser/pressidium-cookie-consent/#trunk)) used and seems to be good!

We probably need to switch to Mozart instead of PhpScoper.

**Comment by mostafasoufi** [comment]
I'll try to finish it next two weeks. @ideenmacher 

Best

**Comment by mostafasoufi** [comment]
Thank you for following up. Our team has been busy rolling out new features, such as Event and Author Tracking. We will promptly address the issue with our development team.

**Comment by mostafasoufi** [comment]
Hey Folks,

We are excited to announce that the issue has been fixed and will be included in version 14.9! You can check it out here: [GitHub Release 14.9](https://github.com/wp-statistics/wp-statistics/tree/release-14.9). This update, coming soon, will bring lots of amazing new features and improvements.

Cheers!
